File: app/src/main/java/com/github/premnirmal/ticker/portfolio/GraphActivity.java
package com.github.premnirmal.ticker.portfolio;

import android.content.DialogInterface;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.github.premnirmal.ticker.BaseActivity;
import com.github.premnirmal.ticker.Injector;
import com.github.premnirmal.ticker.Tools;
import com.github.premnirmal.ticker.model.IHistoryProvider;
import com.github.premnirmal.ticker.model.Range;
import com.github.premnirmal.ticker.model.SerializableDataPoint;
import com.github.premnirmal.ticker.network.Stock;
import com.github.premnirmal.tickerwidget.R;
import com.jjoe64.graphview.GraphView;
import com.jjoe64.graphview.GridLabelRenderer;
import com.jjoe64.graphview.Viewport;
import com.jjoe64.graphview.helper.DateAsXAxisLabelFormatter;
import com.jjoe64.graphview.series.DataPoint;
import com.jjoe64.graphview.series.DataPointInterface;
import com.jjoe64.graphview.series.LineGraphSeries;
import com.jjoe64.graphview.series.OnDataPointTapListener;
import com.jjoe64.graphview.series.PointsGraphSeries;
import com.jjoe64.graphview.series.Series;

import org.joda.time.DateTime;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

import javax.inject.Inject;

import rx.Observable;
import rx.Subscriber;

/**
 * Created by premnirmal on 12/30/14.
 */
public class GraphActivity extends BaseActivity {

    public static final String GRAPH_DATA = "GRAPH_DATA";
    private static final String DATAPOINTS = "DATAPOINTS";
    private static final String RANGE = "RANGE";

    private final DateTimeFormatter formatter = DateTimeFormat.forPattern("MM/dd/YYYY");

    private Stock ticker;
    private SerializableDataPoint[] dataPoints;
    private Range range = Range.THREE_MONTH;

    @Inject
    IHistoryProvider historyProvider;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Injector.inject(this);
        setContentView(R.layout.activity_graph);
        getSupportActionBar().hide();

        ticker = (Stock) getIntent().getSerializableExtra(GRAPH_DATA);
        if (savedInstanceState != null) {
            dataPoints = (SerializableDataPoint[]) savedInstanceState.getSerializable(DATAPOINTS);
            range = (Range) savedInstanceState.getSerializable(RANGE);
        }

        final int viewId;
        switch (range) {
            case ONE_MONTH:
                viewId = R.id.one_month;
                break;
            case THREE_MONTH:
                viewId = R.id.three_month;
                break;
            default:
            case ONE_YEAR:
                viewId = R.id.one_year;
                break;
        }
        findViewById(viewId).setEnabled(false);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putSerializable(DATAPOINTS, dataPoints);
        outState.putSerializable(RANGE, range);
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (dataPoints == null) {
            getData();
        } else {
            loadGraph();
        }
    }

    private void getData() {
        if (Tools.isNetworkOnline(this)) {
            findViewById(R.id.graph_holder).setVisibility(View.GONE);
            findViewById(R.id.progress).setVisibility(View.VISIBLE);
            final Observable<SerializableDataPoint[]> observable = historyProvider.getDataPoints(ticker.symbol, range);
            bind(observable)
                    .subscribe(new Subscriber<SerializableDataPoint[]>() {
                        @Override
                        public void onCompleted() {

                        }

                        @Override
                        public void onError(Throwable e) {
                            showDialog("Error loading datapoints", new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    finish();
                                }
                            });
                        }

                        @Override
                        public void onNext(SerializableDataPoint[] data) {
                            dataPoints = data;
                            loadGraph();
                        }
                    });
        } else {
            showDialog(getString(R.string.no_network_message), new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    finish();
                }
            });
        }
    }

    private void loadGraph() {
        findViewById(R.id.graph_holder).setVisibility(View.VISIBLE);
        findViewById(R.id.progress).setVisibility(View.GONE);
        final GraphView graphView = (GraphView) findViewById(R.id.graph);
        graphView.removeAllSeries();

        final TextView tickerName = (TextView) findViewById(R.id.ticker);
        final TextView desc = (TextView) findViewById(R.id.desc);
        tickerName.setText(ticker.symbol);
        desc.setText(ticker.Name);
        final TextView dataPointValue = (TextView) findViewById(R.id.dataPointValue);
        final LineGraphSeries<DataPoint> series = new LineGraphSeries(dataPoints);
        graphView.addSeries(series);

        final PointsGraphSeries disposableSeries = new PointsGraphSeries(new DataPointInterface[]{dataPoints[dataPoints.length - 1]});
        graphView.addSeries(disposableSeries);
        disposableSeries.setColor(getResources().getColor(R.color.spicy_salmon));
        disposableSeries.setShape(PointsGraphSeries.Shape.POINT);
        disposableSeries.setSize(10f);

        series.setDrawBackground(true);
        series.setBackgroundColor(getResources().getColor(R.color.color_accent));
        series.setOnDataPointTapListener(new OnDataPointTapListener() {
            @Override
            public void onTap(Series series, DataPointInterface dataPointInterface) {
                final StringBuilder dataPointText = new StringBuilder();
                final DateTime dateTime = new DateTime((long) dataPointInterface.getX());
                dataPointText.append(formatter.print(dateTime));
                dataPointText.append(" // ");
                dataPointText.append("$");
                dataPointText.append(dataPointInterface.getY());
                dataPointValue.setText(dataPointText.toString());
                disposableSeries.resetData(new DataPointInterface[]{dataPointInterface});
            }
        });
        final GridLabelRenderer gridLabelRenderer = graphView.getGridLabelRenderer();
        gridLabelRenderer.setLabelFormatter(new DateAsXAxisLabelFormatter(this));
        gridLabelRenderer.setNumHorizontalLabels(5);

        final Viewport viewport = graphView.getViewport();
        viewport.setXAxisBoundsManual(true);
        viewport.setYAxisBoundsManual(true);

        viewport.setMinX(dataPoints[0].getX());
        viewport.setMaxX(dataPoints[dataPoints.length - 1].getX());

        double min = Integer.MAX_VALUE, max = Integer.MIN_VALUE;
        for (int i = 0; i < dataPoints.length; i++) {
            final DataPoint point = dataPoints[i];
            final double val = point.getY();
            if (val < min) {
                min = val;
            } else if (val > max) {
                max = val;
            }
        }
        if (min != Integer.MAX_VALUE && max != Integer.MIN_VALUE) {
            min = min - Math.abs(0.1 * min);
            viewport.setMinY(min <= 0 ? 0 : min);
            viewport.setMaxY(max + Math.abs(0.1 * max));
        }
    }

    /**
     * xml OnClick
     *
     * @param v
     */
    public void updateRange(View v) {
        switch (v.getId()) {
            case R.id.one_month:
                range = Range.ONE_MONTH;
                break;
            case R.id.three_month:
                range = Range.THREE_MONTH;
                break;
            default:
            case R.id.one_year:
                range = Range.ONE_YEAR;
                break;
        }
        final ViewGroup parent = (ViewGroup) v.getParent();
        for (int i = 0; i < parent.getChildCount(); i++) {
            final View view = parent.getChildAt(i);
            if (view != v) {
                view.setEnabled(true);
            } else {
                view.setEnabled(false);
            }
        }
        getData();
    }

}

File: app/src/main/kotlin/com/github/premnirmal/ticker/portfolio/GraphActivity.kt
package com.github.premnirmal.ticker.portfolio

import android.content.DialogInterface
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import com.github.mikephil.charting.animation.Easing
import com.github.mikephil.charting.components.XAxis
import com.github.mikephil.charting.components.YAxis
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import com.github.mikephil.charting.interfaces.datasets.ILineDataSet
import com.github.premnirmal.ticker.Analytics
import com.github.premnirmal.ticker.BaseActivity
import com.github.premnirmal.ticker.CrashLogger
import com.github.premnirmal.ticker.Injector
import com.github.premnirmal.ticker.SimpleSubscriber
import com.github.premnirmal.ticker.Tools
import com.github.premnirmal.ticker.model.IHistoryProvider
import com.github.premnirmal.ticker.model.Range
import com.github.premnirmal.ticker.model.SerializableDataPoint
import com.github.premnirmal.ticker.network.data.Stock
import com.github.premnirmal.tickerwidget.R
import kotlinx.android.synthetic.main.activity_graph.desc
import kotlinx.android.synthetic.main.activity_graph.graphActivityRoot
import kotlinx.android.synthetic.main.activity_graph.graphView
import kotlinx.android.synthetic.main.activity_graph.graph_holder
import kotlinx.android.synthetic.main.activity_graph.progress
import kotlinx.android.synthetic.main.activity_graph.tickerName
import org.threeten.bp.format.DateTimeFormatter
import java.util.ArrayList
import javax.inject.Inject

/**
 * Created by premnirmal on 2/25/16.
 */
class GraphActivity : BaseActivity() {

  private val formatter = DateTimeFormatter.ofPattern("MM/dd/YYYY")

  lateinit private var ticker: Stock
  private var dataPoints: Array<SerializableDataPoint?>? = null
  private var range = Range.THREE_MONTH

  @Inject
  lateinit internal var historyProvider: IHistoryProvider

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    Injector.inject(this)
    setContentView(R.layout.activity_graph)
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
      graphActivityRoot.setPadding(graphActivityRoot.paddingLeft, Tools.getStatusBarHeight(this),
          graphActivityRoot.paddingRight, graphActivityRoot.paddingBottom)
    }
    graphView.isDoubleTapToZoomEnabled = false
    graphView.axisLeft.setDrawGridLines(false)
    graphView.axisLeft.setDrawAxisLine(false)
    graphView.axisLeft.isEnabled = false
    graphView.axisRight.setDrawGridLines(false)
    graphView.axisRight.setDrawAxisLine(true)
    graphView.axisRight.isEnabled = true
    graphView.xAxis.setDrawGridLines(false)
    graphView.setDescription("")
    graphView.legend.isEnabled = false
    graphView.markerView = TextMarkerView(this, graphView)
    if (Build.VERSION.SDK_INT < 16) {
      window.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
          WindowManager.LayoutParams.FLAG_FULLSCREEN)
    } else {
      val decorView = window.decorView
      decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_FULLSCREEN
    }
    ticker = intent.getSerializableExtra(GRAPH_DATA) as Stock
    if (savedInstanceState != null) {
      dataPoints = savedInstanceState.getSerializable(DATAPOINTS) as Array<SerializableDataPoint?>
      range = savedInstanceState.getSerializable(RANGE) as Range
    }

    val viewId: Int
    when (range) {
      Range.ONE_MONTH -> viewId = R.id.one_month
      Range.THREE_MONTH -> viewId = R.id.three_month
      Range.ONE_YEAR -> viewId = R.id.one_year
    }
    findViewById(viewId).isEnabled = false
    Analytics.trackUI("GraphView", ticker.symbol)
  }

  override fun onSaveInstanceState(outState: Bundle) {
    super.onSaveInstanceState(outState)
    outState.putSerializable(DATAPOINTS, dataPoints)
    outState.putSerializable(RANGE, range)
  }

  override fun onResume() {
    super.onResume()
    if (dataPoints == null) {
      getData()
    } else {
      loadGraph(dataPoints!!)
    }
  }

  private fun getData() {
    if (Tools.isNetworkOnline(this)) {
      findViewById(R.id.graph_holder).visibility = View.GONE
      findViewById(R.id.progress).visibility = View.VISIBLE
      val observable = historyProvider.getDataPoints(ticker.symbol, range)
      bind(observable).subscribe(object : SimpleSubscriber<Array<SerializableDataPoint?>>() {
        override fun onError(e: Throwable) {
          CrashLogger.logException(e)
          showDialog("Error loading datapoints",
              DialogInterface.OnClickListener { dialog, which -> finish() })
        }

        override fun onNext(result: Array<SerializableDataPoint?>) {
          dataPoints = result
          loadGraph(dataPoints!!)
        }
      })
    } else {
      showDialog(getString(R.string.no_network_message),
          DialogInterface.OnClickListener { dialog, which -> finish() })
    }
  }

  private fun loadGraph(points: Array<SerializableDataPoint?>) {
    graphView.lineData?.clearValues()
    graphView.invalidate()
    val dataPointsList = points.toList()
    tickerName.text = ticker.symbol
    desc.text = ticker.Name
    val series = LineDataSet(dataPointsList.toList(), range.name)
    series.setDrawHorizontalHighlightIndicator(false)
    series.setDrawValues(false)
    val color_accent = resources.getColor(R.color.color_accent)
    series.setDrawFilled(true)
    series.color = color_accent
    series.fillColor = color_accent
    series.fillAlpha = 150
    series.setDrawCubic(true)
    series.cubicIntensity = 0.07f
    series.lineWidth = 2f
    series.setDrawCircles(false)
    series.highLightColor = Color.GRAY
    val dataSets: MutableList<ILineDataSet> = ArrayList()
    dataSets.add(series)
    val xDataSet: MutableList<String> = dataPointsList.indices.mapTo(
        ArrayList()) { formatter.format(dataPointsList[it]?.getQuote()?.date) }
    val lineData: LineData = LineData(xDataSet, dataSets)
    graphView.data = lineData
    val xAxis: XAxis = graphView.xAxis
    val yAxis: YAxis = graphView.axisRight
    xAxis.position = XAxis.XAxisPosition.BOTTOM
    xAxis.textSize = 10f
    yAxis.textSize = 10f
    xAxis.textColor = Color.GRAY
    yAxis.textColor = Color.GRAY
    xAxis.setLabelsToSkip(xDataSet.size / 5)
    yAxis.setPosition(YAxis.YAxisLabelPosition.OUTSIDE_CHART)
    xAxis.setDrawAxisLine(true)
    yAxis.setDrawAxisLine(true)
    xAxis.setDrawGridLines(false)
    yAxis.setDrawGridLines(false)
    graph_holder.visibility = View.VISIBLE
    progress.visibility = View.GONE
    graphView.animateX(DURATION, Easing.EasingOption.EaseInOutQuad)
  }

  /**
   * xml OnClick

   * @param v
   */
  fun updateRange(v: View) {
    when (v.id) {
      R.id.one_month -> range = Range.ONE_MONTH
      R.id.three_month -> range = Range.THREE_MONTH
      R.id.one_year -> range = Range.ONE_YEAR
    }
    Analytics.trackUI("GraphUpdateRange", ticker.symbol + "-" + range.name)
    val parent = v.parent as ViewGroup
    (0..parent.childCount - 1)
        .map { parent.getChildAt(it) }
        .forEach { it.isEnabled = it != v }
    getData()
  }

  companion object {

    val GRAPH_DATA = "GRAPH_DATA"
    private val DATAPOINTS = "DATAPOINTS"
    private val RANGE = "RANGE"
    private val DURATION = 2000
  }

}